library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java

```diff
@@ -106,94 +106,95 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-            waitingForConnectionToRequestLocationUpdate(false);
-            locationRequired();
-        } else {
-            LogUtils.logI("We got location, no need to ask for location updates.");
-        }
-    }
 
-    @Override
-    public void onConnectionSuspended(int i) {
-        if (!getConfiguration().googlePlayServicesConfiguration().failOnConnectionSuspended()
-              && suspendedConnectionIteration < getConfiguration().googlePlayServicesConfiguration()
-              .suspendedConnectionRetryCount()) {
-            LogUtils.logI("GoogleApiClient connection is suspended, try to connect again.");
-            suspendedConnectionIteration++;
-            getSourceProvider().connectGoogleApiClient();
+            // Request fresh location
+            requestLocation(false);
         } else {
-            LogUtils.logI("GoogleApiClient connection is suspended, calling fail...");
-            failed(FailType.GOOGLE_PLAY_SERVICES_CONNECTION_FAIL);
+            // Try to get last location, if failed then request fresh location
+            checkLastKnowLocation();
         }
     }
 
-    @Override
-    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
-        LogUtils.logI("GoogleApiClient connection is failed.");
-        failed(FailType.GOOGLE_PLAY_SERVICES_CONNECTION_FAIL);
-    }
-
-    @Override
-    public void onLocationChanged(Location location) {
+    public void onLocationChanged(@NonNull Location location) {
         if (getListener() != null) {
             getListener().onLocationChanged(location);
         }
 
         // Set waiting as false because we got at least one, even though we keep tracking user's location
         setWaiting(false);
 
-        if (!getConfiguration().keepTracking() && getSourceProvider().isGoogleApiClientConnected()) {
+        if (!getConfiguration().keepTracking()) {
+            // If need to update location once, clear the listener to prevent multiple call
             LogUtils.logI("We got location and no need to keep tracking, so location update is removed.");
-            getSourceProvider().removeLocationUpdates();
+
+            removeLocationUpdates();
         }
     }
 
     @Override
-    public void onResult(@NonNull LocationSettingsResult result) {
-        final Status status = result.getStatus();
-
-        switch (status.getStatusCode()) {
-            case LocationSettingsStatusCodes.SUCCESS:
-                // All location settings are satisfied. The client can initialize location
-                // requests here.
-                LogUtils.logI("We got GPS, Wifi and/or Cell network providers enabled enough "
-                      + "to receive location as we needed. Requesting location update...");
-                requestLocationUpdate();
-                break;
+    public void onLocationResult(@Nullable LocationResult locationResult) {
+        if (locationResult == null) {
+            // Do nothing, wait for other result
```

Comment:
Is that a thing? Are you sure we shouldn't be failing here?

Reply:
As I can understand in obfuscated code, `locationResult` can be null if the first number in Parcel var0 header is 0:
```
// Called from here com.google.android.gms.location.zzv.dispatchTransaction()
// Class com.google.android.gms.internal.location.zzc
public static <T extends Parcelable> T zza(Parcel var0, Creator<T> var1) {
        return var0.readInt() == 0 ? null : (Parcelable)var1.createFromParcel(var0);
 }
```
I did'n find cause when first Int can be 0. Maybe in service disconnection phase all received LocationResults sets as unwanted?

Reply:
Yes, that is fine, if they think API is nullable, then we should treat that way for sure.
My question was about "wait for other result" is that a thing? Is this a continuous task that calls the listener multiple times or will it terminate once it is called? If it will terminate, we need to terminate here as well, and potentially with an error.

Reply:
This is a continuous task, the connection will be active until removeLocationUpdates() is called.

Reply:
I see, thanks for the explanation.